server/src/main/java/io/crate/planner/operators/LogicalPlanner.java

```diff
@@ -647,6 +665,34 @@ public static void executeNodeOpTree(DependencyCarrier dependencies,
             .execute(consumer, txnCtx);
     }
 
+    private class MultiPhaseOptimizer extends LogicalPlanVisitor<PlannerContext, LogicalPlan> {
```

Comment:
I had more in mind to do it here: 

https://github.com/crate/crate/blob/e668e759588f869a66d4c710d4a6c631b51ec78b/server/src/main/java/io/crate/execution/MultiPhaseExecutor.java#L51

That would have the advantage that the plan/sources doesn't need to be copied